gobblin-core/src/main/java/org/apache/gobblin/converter/avro/JsonRecordAvroSchemaToAvroConverter.java

```diff
@@ -0,0 +1,92 @@
+import java.util.List;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.gobblin.configuration.WorkUnitState;
+import org.apache.gobblin.converter.Converter;
+import org.apache.gobblin.converter.DataConversionException;
+import org.apache.gobblin.converter.SchemaConversionException;
+import org.apache.gobblin.converter.SingleRecordIterable;
+import org.apache.gobblin.converter.ToAvroConverterBase;
+import com.google.gson.JsonObject;
+
+
+/**
+ * {@link Converter} that takes an Avro schema in string format and corresponding {@link JsonObject} records and
+ * converts them to {@link GenericRecord} using the schema
+ */
+public class JsonRecordAvroSchemaToAvroConverter extends ToAvroConverterBase<String, JsonObject> {
+
+  /**
+   * Take an Avro schema in string format and parse it into a {@link Schema}
+   */
+  @Override
+  public Schema convertSchema(String schema, WorkUnitState workUnit) throws SchemaConversionException {
+    return new Schema.Parser().parse(schema);
+  }
+
+  /**
+   * Take in {@link JsonObject} input records and convert them to {@link GenericRecord} using outputSchema
+   */
+  @Override
+  public Iterable<GenericRecord> convertRecord(Schema outputSchema, JsonObject inputRecord, WorkUnitState workUnit)
+      throws DataConversionException {
+    GenericRecord avroRecord = convertNestedRecord(outputSchema, inputRecord, workUnit);
+    return new SingleRecordIterable<>(avroRecord);
+  }
+
+  private GenericRecord convertNestedRecord(Schema outputSchema, JsonObject inputRecord, WorkUnitState workUnit) throws DataConversionException {
+    GenericRecord avroRecord = new GenericData.Record(outputSchema);
+    JsonElementConversionWithAvroSchemaFactory.JsonElementConverter converter;
+    for (Schema.Field field : outputSchema.getFields()) {
+      Schema.Type type = field.schema().getType();
+      if (type.equals(Schema.Type.RECORD)) {
+        avroRecord.put(field.name(), convertNestedRecord(field.schema(), inputRecord.get(field.name()).getAsJsonObject(), workUnit));
+      } else {
+        boolean nullable = false;
+
+        if (type.equals(Schema.Type.UNION)) {
+          nullable = true;
+          List<Schema> types = field.schema().getTypes();
+          if (types.size() != 2) {
+            throw new DataConversionException("Only unions of size 2 supported");
+          }
+          if (field.schema().getTypes().get(0).getType().equals(Schema.Type.NULL)) {
+            type = field.schema().getTypes().get(1).getType();
+          } else if (field.schema().getTypes().get(1).getType().equals(Schema.Type.NULL)) {
+            type = field.schema().getTypes().get(0).getType();
+          } else {
+            throw new DataConversionException("Union must contain null");
+          }
+        }
+        try {
+          converter = JsonElementConversionWithAvroSchemaFactory.getConvertor(field.name(), type.getName(), field.schema(),
```

Comment:
the type extracted from UNION at line 74 or 76 can be a RECORD (eg. UNION(null, RECORD) ), so I think you should handle that by invoking nested conversion as you did for RECORD earlier?